### PR TITLE
runners: add support for stm32cubeprogrammer 

### PR DESCRIPTION
**Pull request by gmarull**
Commit summary:
- Add support for the official ST Microelectronics programming tool (CLI version).
- Updated NUCLEO-H743ZI board to support the STM32CubeProgrammer runner.

### CODE REVIEW

File: scripts/west_commands/runners/stm32cubeprogrammer.py
Comment (gmarull) [inline scripts/west_commands/runners/stm32cubeprogrammer.py:63]: @mbolivar-nordic would be nice to allow args/kwargs on `check_call`, so that environment can be set using `env` on that call.

File: scripts/west_commands/runners/stm32cubeprogrammer.py
Comment (ABOSTM) [inline scripts/west_commands/runners/stm32cubeprogrammer.py:62]: @gmarull,
I am working on windows, so I can test if you want.
But I would like to be sure code is ready for windows? (there is no else in `if sys.platform == "linux"`)

File: scripts/west_commands/runners/stm32cubeprogrammer.py
Comment (ABOSTM) [inline scripts/west_commands/runners/stm32cubeprogrammer.py:75]: My cubeProgrammer instalation in windows is there:
c:\Program Files\STMicroelectronics\STM32Cube\STM32CubeProgrammer\
it is not in c:\Program Files (x86)\

Comment (gmarull) [reply scripts/west_commands/runners/stm32cubeprogrammer.py]: Interesting, mine is at x86 folder. Can you determine what is the expected path? Maybe new versions install on non-x86? Are you on a 64-bit system?

File: scripts/west_commands/runners/stm32cubeprogrammer.py
Comment (gmarull) [reply scripts/west_commands/runners/stm32cubeprogrammer.py]: I think this is only a requirement for Linux, but would be nice to confirm.

File: scripts/west_commands/runners/stm32cubeprogrammer.py
Comment (mbolivar-nordic) [reply scripts/west_commands/runners/stm32cubeprogrammer.py]: > @mbolivar-nordic would be nice to allow args/kwargs on `check_call`, so that environment can be set using `env` on that call.

Sure, feel free to add this in a separate commit in this PR.

File: scripts/west_commands/runners/stm32cubeprogrammer.py
Comment (ABOSTM) [reply scripts/west_commands/runners/stm32cubeprogrammer.py]: Yes, I am on 64-bit system. 
I am not sure to be able to determine THE expected path, it looks like there are severalpaths possible (depending on version installed or system version). So I would propose to check the presence of executable in one path and if not available check the other path.

Comment (gmarull) [reply scripts/west_commands/runners/stm32cubeprogrammer.py]: Could you list all the *official* paths? In case the user decides to install to a custom location the `--cli` option is available.

File: scripts/west_commands/runners/stm32cubeprogrammer.py
Comment (mbolivar-nordic) [inline scripts/west_commands/runners/stm32cubeprogrammer.py:20]: Why is this under TYPE_CHECKING?

It isn't a performance optimization AFAICT. Both argparse and runners.core have already been imported, so the imports in this file will hit the sys.modules cache and return immediately. Is the linter complaining without the `if` somehow?

File: scripts/west_commands/runners/stm32cubeprogrammer.py
Comment (mbolivar-nordic) [inline scripts/west_commands/runners/stm32cubeprogrammer.py:134]: We have various legacy options in different runners that map directly to an option that is passed to the underlying tool, but I am trying to get away from doing this by adding a single `--tool-opt` option that says "hey runner, here is a thing you can pass to the program you're a wrapper around".

The nrfjprog, jlink, and pyocd runners implement this today. I see this here and think "could this be a --tool-opt instead"?

Are there any other options that you think might have a "yes" answer to this question in this PR?

File: scripts/west_commands/runners/stm32cubeprogrammer.py
Comment (mbolivar-nordic) [inline scripts/west_commands/runners/stm32cubeprogrammer.py:152]: Similarly it feels like this could be renamed --tool-opt for consistency with the other runners.

**Review by mbolivar-nordic — COMMENTED** [review]
I didn't review the implementation, but I have some questions and comments about the command line options, and a random one about TYPE_CHECKING.

File: scripts/west_commands/runners/stm32cubeprogrammer.py
Comment (gmarull) [reply scripts/west_commands/runners/stm32cubeprogrammer.py:20]: Good catch, both will already be imported.

File: scripts/west_commands/runners/stm32cubeprogrammer.py
Comment (gmarull) [reply scripts/west_commands/runners/stm32cubeprogrammer.py:134]: I did not know `--tool-opt` was standardized for passing arbitrary tool options. Should be put in the base class as `--erase` then? The `connect-opts` will be needed, as these are appended options specific to the connection set.

Comment (mbolivar-nordic) [reply scripts/west_commands/runners/stm32cubeprogrammer.py:134]: > I did not know `--tool-opt` was standardized for passing arbitrary tool options. Should it be put in the base class as `--erase` then?

Yeah, eventually it probably should be a common capability; I just haven't gotten around to it yet. I mostly just ask people for this change when I see this sort of option being added.

> The `connect-opts` will be needed, as these are appended options specific to the connection set.

OK, so kind of like --dfuse-modifiers then? Seems reasonable to me.

Comment (gmarull) [reply scripts/west_commands/runners/stm32cubeprogrammer.py]: @mbolivar fixed, used `--conn-modifiers`

File: scripts/west_commands/runners/stm32cubeprogrammer.py
Comment (gmarull) [reply scripts/west_commands/runners/stm32cubeprogrammer.py]: fixed

File: scripts/west_commands/runners/stm32cubeprogrammer.py
Comment (gmarull) [reply scripts/west_commands/runners/stm32cubeprogrammer.py]: fixed

**Review by mbolivar-nordic — APPROVED** [review]
Command line options look fine to me. I leave it up to the STM32 SoC maintainers to give the second approval for the implementation.

**Comment by gmarull** [comment]
@ABOSTM @erwango ping

File: scripts/west_commands/tests/test_stm32cubeprogrammer.py
Comment (ABOSTM) [inline scripts/west_commands/tests/test_stm32cubeprogrammer.py:21]: > Could you list all the official paths? In case the user decides to install to a custom location the --cli option is available.

I sent email but have no response up to now.

What about your path "c:\Program Files (x86)" ?

Comment (gmarull) [reply scripts/west_commands/tests/test_stm32cubeprogrammer.py:21]: I had 2.1.0 installed. Uninstalled and downloaded latest 2.5.0, it still suggests x86 path:

image

I'm on Windows 10 1903 x64

image

Comment (ABOSTM) [reply scripts/west_commands/tests/test_stm32cubeprogrammer.py:21]: > What about your path "c:\Program Files (x86)" ?

I mean:
Why not tackling both known paths : 
"c:\Program Files\STMicroelectronics\STM32Cube\STM32CubeProgrammer"
and "c:\Program Files (x86)\"
(by checking existence of cube programmer in one or both paths ?)

**Review by ABOSTM — APPROVED** [review]
LGTM

**Comment by gmarull** [comment]
@carlescufi docs CI broken? https://github.com/zephyrproject-rtos/zephyr/pull/29102/checks?check_run_id=1342235943

**Review by erwango — APPROVED** [review]

**Review by MaureenHelm — APPROVED** [review]

**Pull request merged**
